src/cwe_checker_lib/src/utils/binary.rs

```diff
@@ -175,15 +175,51 @@ impl RuntimeMemoryImage {
             if address >= segment.base_address
                 && address <= segment.base_address + segment.bytes.len() as u64
             {
-                let index = (address - segment.base_address) as usize;
-                let c_str = std::ffi::CStr::from_bytes_with_nul(&segment.bytes[index..])?;
-                return Ok(c_str.to_str()?);
+                let start_index = (address - segment.base_address) as usize;
+                if let Some(end_index) = segment.bytes[start_index..].iter().position(|&b| b == 0) {
+                    let c_str = std::ffi::CStr::from_bytes_with_nul(
+                        &segment.bytes[start_index..start_index + end_index + 1],
+                    )?;
+                    return Ok(c_str.to_str()?);
+                } else {
+                    return Err(anyhow!("Not a valid string in memory."));
+                }
             }
         }
 
         Err(anyhow!("Address is not a valid global memory address."))
     }
 
+    /// Checks whether the constant is a global memory address.
+    pub fn is_global_memory_address(&self, constant: &Bitvector) -> bool {
+        if self.read(constant, constant.bytesize()).is_ok() {
+            return true;
+        }
+        false
+    }
+
+    /// Checks whether the memory content at the input address is
+    /// an address to another memory position and returns the address in memory.
+    pub fn parse_address_if_recursive(
+        &self,
+        address: &Bitvector,
+        arch_size: ByteSize,
+    ) -> Result<Bitvector, Error> {
+        match self.read(address, arch_size) {
+            Ok(Some(recursive_address)) => {
+                if self.read(&recursive_address, arch_size).is_ok() {
+                    Ok(recursive_address)
+                } else {
+                    Ok(address.clone())
+                }
+            }
+            Ok(None) => Err(anyhow!(
+                "Writeable address does not guarantee correct content."
+            )),
+            Err(_) => Ok(address.clone()),
+        }
+    }
```

Comment:
Should this function really return `Ok(address.clone())` in some cases? It may confuse the caller if the caller expects the result being the target of address and not address itself.

Reply:
The caller always expects an address to be returned if there is no error. The caller will then resolve the target content. This function's purpose is to parse an recursive address if this is the case. So it could either be (address -> address -> content) or (address -> content). In the first case, the second address is returned.

Reply:
But could it happen that in the (address -> content) case the content is misinterpreted as an address?